src/proof_generation/scripts/metamath-extract.py

```diff
@@ -183,18 +184,90 @@ def slice_database(input_database: Database) -> Iterator[tuple[str, Database]]:
+    while ascii[-1] == '=':
+        ascii = ascii[:-1]
+    return ascii
+
+
+abbreviation_index: dict[str, str] = {
+    'goal': 'goal',
+    '#Variable': '#Variable',
+    '#ElementVariable': '#ElementVariable',
+    '#SetVariable': '#SetVariable',
+    '#Pattern': '#Pattern',
+    '#Symbol': '#Symbol',
+}
+
+
+def abbreviate(name: str) -> str:
+    if not name in abbreviation_index:
+        abbreviation_index[name] = int_to_abbreviation(len(abbreviation_index))
+    assert abbreviation_index[name], abbreviation_index
+    return abbreviation_index[name]
+
+
+def abbreviate_constructors(term: Term) -> Term:
+    if isinstance(term, Metavariable):
+        return term
+    elif isinstance(term, Application):
+        return Application(abbreviate(term.symbol), term.subterms)
+    else:
+        raise RuntimeError(f'Unexpected term type, {type(term)}')
+
+
+def terms_abbreviate_constructors(terms: Terms) -> Terms:
+    return tuple(term.bottom_up(abbreviate_constructors) for term in terms)
+
+
+def abbreviate_lemmas(statement: Statement) -> Statement:
+    if not isinstance(statement, StructuredStatement):
+        return statement
+    else:
+        new_label = abbreviate(statement.label)
+
+        if isinstance(statement, FloatingStatement):
+            return FloatingStatement(new_label, statement.terms)
+        elif isinstance(statement, EssentialStatement):
+            return EssentialStatement(new_label, terms_abbreviate_constructors(statement.terms))
+        elif isinstance(statement, AxiomaticStatement):
+            return AxiomaticStatement(new_label, terms_abbreviate_constructors(statement.terms))
+        elif isinstance(statement, ProvableStatement):
+            proof = statement.proof
+            if proof:
+                lemmas: Iterable[str]
+                lemmas, lemma_applications = deconstruct_compressed_proof(proof)
+                lemmas = (abbreviate(lemma) for lemma in lemmas)
+                proof = '( {} ) {}'.format(' '.join(lemmas), lemma_applications)
+            return ProvableStatement(new_label, terms_abbreviate_constructors(statement.terms), proof=proof)
+        else:
+            raise RuntimeError(f'Unexpected statement type, {type(statement)}')
+
+
 def main() -> None:
     parser = argparse.ArgumentParser()
     parser.add_argument('input', help='Input Metamath file')
     parser.add_argument('output', help='Output Metamath directory')
+    parser.add_argument('--abbreviate-lemma-names', action=argparse.BooleanOptionalAction)
     args = parser.parse_args()
 
-    print('Parsing database.')
+    print('Parsing database...', end='')
     input_database = load_database(args.input, include_proof=True)
-    print('Database parsed.')
+    print(' Done.')
 
     output_dir = Path(args.output)
     output_dir.mkdir()
+
+    if args.abbreviate_lemma_names:
+        print('Abbreviating Lemma names...', end='')
+        input_database = input_database.bottom_up(abbreviate_lemmas)
```

Comment:
Since we're reconstructing the databases for the slices anyway, would not it be better to do it together and not in a preprocessing step?

Reply:
There's a couple of reasons:

* I like the separation of concerns.
*  At least from the architectural point of view, there shouldn't be much of a performance penalty---we can throw away the original statements once they are processed. (Note that the current implementation keeps the initial database until the abreviated database is produced though).

Reply:
Right, but it still needs to go over them twice, albeit the second time is over the abbreviated ones... But it's true I like the separation of concerns too.